modules/3d/src/rgbd/odometry.cpp

```diff
@@ -2,2475 +2,343 @@
 }
 
-template<typename TMat>
-static
-void preparePyramidSobel(InputArrayOfArrays pyramidImage, int dx, int dy, InputOutputArrayOfArrays pyramidSobel)
+void OdometryRGB::prepareFrame(OdometryFrame frame)
 {
-    size_t imgLevels = pyramidImage.size(-1).width;
-    size_t sobelLvls = pyramidSobel.size(-1).width;
-    if(!pyramidSobel.empty())
-    {
-        if(sobelLvls != imgLevels)
-            CV_Error(Error::StsBadSize, "Incorrect size of pyramidSobel.");
-
-        for(size_t i = 0; i < sobelLvls; i++)
-        {
-            CV_Assert(pyramidSobel.size((int)i) == pyramidImage.size((int)i));
-            CV_Assert(pyramidSobel.type((int)i) == CV_16SC1);
-        }
-    }
-    else
-    {
-        pyramidSobel.create((int)imgLevels, 1, CV_16SC1, -1);
-        for(size_t i = 0; i < imgLevels; i++)
-        {
-            Sobel(getTMat<TMat>(pyramidImage, (int)i), getTMatRef<TMat>(pyramidSobel, (int)i), CV_16S, dx, dy, sobelSize);
-        }
-    }
+    prepareRGBFrame(frame, frame, this->settings);
 }
 
-static
-void randomSubsetOfMask(InputOutputArray _mask, float part)
+void OdometryRGB::prepareFrames(OdometryFrame srcFrame, OdometryFrame dstFrame)
 {
-    const int minPointsCount = 1000; // minimum point count (we can process them fast)
-    const int nonzeros = countNonZero(_mask);
-    const int needCount = std::max(minPointsCount, int(_mask.total() * part));
-    if(needCount < nonzeros)
-    {
-        RNG rng;
-        Mat mask = _mask.getMat();
-        Mat subset(mask.size(), CV_8UC1, Scalar(0));
-
-        int subsetSize = 0;
-        while(subsetSize < needCount)
-        {
-            int y = rng(mask.rows);
-            int x = rng(mask.cols);
-            if(mask.at<uchar>(y,x))
-            {
-                subset.at<uchar>(y,x) = 255;
-                mask.at<uchar>(y,x) = 0;
-                subsetSize++;
-            }
-        }
-        _mask.assign(subset);
-    }
+    prepareRGBFrame(srcFrame, dstFrame, this->settings);
 }
 
-static
-void preparePyramidTexturedMask(InputArrayOfArrays pyramid_dI_dx, InputArrayOfArrays pyramid_dI_dy,
-                                InputArray minGradMagnitudes, InputArrayOfArrays pyramidMask, double maxPointsPart,
-                                InputOutputArrayOfArrays pyramidTexturedMask)
+bool OdometryRGB::compute(OdometryFrame srcFrame, OdometryFrame dstFrame, OutputArray Rt) const
 {
-    size_t didxLevels = pyramid_dI_dx.size(-1).width;
-    size_t texLevels = pyramidTexturedMask.size(-1).width;
-    if(!pyramidTexturedMask.empty())
-    {
-        if(texLevels != didxLevels)
-            CV_Error(Error::StsBadSize, "Incorrect size of pyramidTexturedMask.");
+    Matx33f cameraMatrix;
+    settings.getCameraMatrix(cameraMatrix);
+    std::vector<int> iterCounts;
+    Mat miterCounts;
+    settings.getIterCounts(miterCounts);
+    for (int i = 0; i < miterCounts.size().height; i++)
+        iterCounts.push_back(miterCounts.at<int>(i));
```

Comment:
> `miterCounts.at<int>`

Before using of low-level access we should check matrix type through `CV_Assert` / `CV_CheckTypeEQ`

Reply:
done

Reply:
@DumDereDum this should be checked before `miterCounts` accessing, not after

Reply:
done